x/merkledb/intermediate_node_db.go

```diff
@@ -108,17 +108,43 @@ func (db *intermediateNodeDB) Get(key path) (*node, error) {
 	}
 	db.metrics.IntermediateNodeCacheMiss()
 
-	prefixedKey := addPrefixToKey(db.bufferPool, intermediateNodePrefix, key.Bytes())
+	dbKey := db.constructDBKey(key)
 	db.metrics.DatabaseNodeRead()
-	nodeBytes, err := db.baseDB.Get(prefixedKey)
+	nodeBytes, err := db.baseDB.Get(dbKey)
 	if err != nil {
 		return nil, err
 	}
-	db.bufferPool.Put(prefixedKey)
+	db.bufferPool.Put(dbKey)
 
 	return parseNode(key, nodeBytes)
 }
 
+// constructDBKey creates a key that can be used in the base db
+// We need to be able to differentiate between two paths of equal byte length but different token length,
+// so it adds a padding that can be used to differentiate.
+// Additionally, it adds the prefix indicating it is part of the intermediateDB
+func (db *intermediateNodeDB) constructDBKey(key path) []byte {
+	compressedKey := key.Serialize()
+
+	// add one additional byte to store padding when the path has a length that fits into a whole number of bytes
+	remainder := compressedKey.NibbleLength % 2
+	keyLen := len(compressedKey.Value)
+	if remainder == 0 {
+		keyLen++
+	}
+	dbKey := getBufferFromPool(db.bufferPool, keyLen)
+	defer db.bufferPool.Put(dbKey)
+
+	copy(dbKey, compressedKey.Value)
+	if remainder == 0 {
+		dbKey[keyLen-1] = 0b1000_0000
+	} else {
+		dbKey[keyLen-1] += 0b0000_1000
```

Comment:
Personally - I feel like this would be more clear as a bitwise OR rather than an addition. But both should work.